rotkehlchen/exchanges/kucoin.py

```diff
@@ -0,0 +1,775 @@
+            case: Literal['validate_api_key', 'balances', 'trades', 'asset_movements'],
+    ) -> Union[
+        List,
+        Tuple[bool, str],
+        ExchangeQueryBalances,
+    ]:
+        """Process unsuccessful responses
+
+        May raise RemoteError and SystemClockNotSyncedError
+        """
+        try:
+            response_dict = rlk_jsonloads_dict(response.text)
+        except JSONDecodeError as e:
+            msg = f'Kucoin {case} returned an invalid JSON response: {response.text}.'
+            log.error(msg)
+
+            if case in ('validate_api_key', 'balances'):
+                raise RemoteError(msg) from e
+            if case in ('trades', 'asset_movements'):
+                self.msg_aggregator.add_error(
+                    f'Got remote error while querying Kucoin {case}: {msg}',
+                )
+                return []
+
+            raise AssertionError(f'Unexpected case: {case}') from e
+
+        error_code = response_dict.get('code', None)
+        self._check_for_system_clock_not_synced_error(error_code)
+        # Errors related with the API key return a human readable message
+        if case == 'validate_api_key' and error_code in API_KEY_ERROR_CODE_ACTION.keys():
+            return False, API_KEY_ERROR_CODE_ACTION[response_dict['code']]
+
+        # Below any other error not related with the system clock or the API key
+        reason = response_dict.get('msg', None) or response.text
+        msg = (
+            f'Kucoin query responded with error status code: {response.status_code} '
+            f'and text: {reason}.'
+        )
+        log.error(msg)
+
+        if case == 'validate_api_key':
+            raise RemoteError(msg)
+        if case == 'balances':
+            return None, msg
+        if case in ('trades', 'asset_movements'):
+            self.msg_aggregator.add_error(
+                f'Got remote error while querying Kucoin {case}: {msg}',
+            )
+            return []
+
+        raise AssertionError(f'Unexpected case: {case}')
+
+    def first_connection(self) -> None:
+        self.first_connection_made = True
+
+    @protect_with_lock()
+    @cache_response_timewise()
+    def query_balances(self) -> ExchangeQueryBalances:
+        """Return the account balances
+
+        May raise RemoteError and SystemClockNotSyncedError
+        """
+        accounts_response = self._api_query('accounts')
+        if accounts_response.status_code != HTTPStatus.OK:
+            result, msg = self._process_unsuccessful_response(
+                response=accounts_response,
+                case='balances',
+            )
+            return result, msg
+
+        try:
+            response_dict = rlk_jsonloads_dict(accounts_response.text)
+        except JSONDecodeError as e:
+            msg = f'Kucoin balances returned an invalid JSON response: {accounts_response.text}.'
+            log.error(msg)
+            raise RemoteError(msg) from e
+
+        try:
```

Comment:
Also this try/except I would except it to be part of `self._calculate_accounts_balances`, which is more like `self._deserialize_accounts_balances`